### ISSUE TITLE
 64-bit architecture

### ISSUE THREAD

**Issue by bencz** [ELENA-LANG/elena-lang#206]
I was doing some analysis on start support to x64 in asm2binx, and I thought of creating a file only to the registers and a helper only for x64.
On separate registers to a file, in which the x64 were added several registers.

| 64-bit register | Lower 32 bits | Lower 16 bits | Lower 8 bits |
| --- | --- | --- | --- |
| rax | eax | ax | al |
| rbx | ebx | bx | bl |
| rcx | ecx | cx | cl |
| rdx | edx | dx | dl |
| rsi | esi | si | sil |
| rdi | edi | di | dil |
| rbp | ebp | bp | bpl |
| rsp | esp | sp | spl |
| r8 | r8d | r8w | r8b |
| r9 | r9d | r9w | r9b |
| r10 | r10d | r10w | r10b |
| r11 | r11d | r11w | r11b |
| r12 | r12d | r12w | r12b |
| r13 | r13d | r13w | r13b |
| r14 | r14d | r14w | r14b |
| r15 | r15d | r15w | r15b |

The calculation of dislocation/register is exactly the same for x86, the difference is than in x64 use a prefix to know if access x64 or not.
For example:
48 89 d8  -- mov rax, rbx
The mov opcode instruction for x64 is: REX.W + 89 /r    --- MOV r/m64,r64 
( http://www.tptp.cc/mirrors/siyobik.info/instruction/MOV.html )

About REX., look at http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-manual-325462.pdf, in page 81
http://wiki.osdev.org/X86-64_Instruction_Encoding#REX_prefix

About the assemble process, maybe is more better create a 
- assembler.cpp
- x64assembler.cpp

The assembler file will contains the controller of assemble process.

**Comment by bencz** [comment]
Some interesting links:
http://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64/
https://msdn.microsoft.com/en-us/library/ew5tede7.aspx

**Comment by arakov** [comment]
thanks for links. yes there should be x86assembler and x64assembler modules

**Comment by bencz** [comment]
Hi :)
About assembler, I mean, rename the x86assembler.cpp to assembler.cpp, create a x64helper, and add in assembly.cpp, instructions, support for x64 :)

**Comment by arakov** [comment]
Why? There should be clear difference between x86 and x64. So if there will be common code it should stay in assembler.h

**Comment by bencz** [comment]
Hi
what I'm thinking, it's about the compilation of instructions...
mov eax, ebx 
mov rax, rbx 

I am imagining, is just add support for 64bit registers on existing methods of instructions compilation..
See what I mean?

**Comment by arakov** [comment]
Yes, I see. but I'm not sure if it could be so easily extended. We will see

**Comment by arakov** [comment]
ELENA 64 will not maintain 16 byte stack alignment due to the current call conventions and some language features (like a command tape). The alignment will be maintained only for external operations
